core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinFastRecordCursorFactory.java

```diff
@@ -0,0 +1,460 @@
+        private long resize(long entrySize, long appendAddress) {
+            assert appendAddress >= heapStart;
+            long currentCapacity = heapLimit - heapStart;
+            long newCapacity = currentCapacity << 1;
+            long target = appendAddress + entrySize - heapStart;
+            if (newCapacity < target) {
+                newCapacity = Numbers.ceilPow2(target);
+            }
+            if (newCapacity > MAX_HEAP_SIZE) {
+                throw LimitOverflowException.instance().put("limit of ").put(MAX_HEAP_SIZE).put(" memory exceeded in ASOF join");
+            }
+
+            long kAddress = Unsafe.realloc(heapStart, currentCapacity, newCapacity, MemoryTag.NATIVE_JOIN_MAP);
+
+            long delta = kAddress - heapStart;
+            this.heapStart = kAddress;
+            this.heapLimit = kAddress + newCapacity;
+
+            return delta;
+        }
+
+        protected void checkCapacity(long requiredSize) {
+            if (appendAddress + requiredSize > heapLimit) {
+                long delta = resize(requiredSize, appendAddress);
+                heapStart += delta;
+                appendAddress += delta;
+                assert heapStart > 0;
+                assert appendAddress > 0;
+            }
+        }
+    }
+
+    private class AsOfJoinKeyedFastRecordCursor extends AbstractAsOfJoinFastRecordCursor {
+        private int origSlaveFrameIndex = -1;
+        private long origSlaveRowId = -1;
+
+        public AsOfJoinKeyedFastRecordCursor(
+                int columnSplit,
+                Record nullRecord,
+                int masterTimestampIndex,
+                int slaveTimestampIndex,
+                int lookahead
+        ) {
+            super(columnSplit, nullRecord, masterTimestampIndex, slaveTimestampIndex, lookahead);
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (isMasterHasNextPending) {
+                masterHasNext = masterCursor.hasNext();
+                isMasterHasNextPending = false;
+            }
+            if (!masterHasNext) {
+                return false;
+            }
+
+            if (origSlaveRowId != -1) {
+                slaveCursor.recordAt(slaveRecB, Rows.toRowID(origSlaveFrameIndex, origSlaveRowId));
+            }
+            final long masterTimestamp = masterRecord.getTimestamp(masterTimestampIndex);
+            if (masterTimestamp >= lookaheadTimestamp) {
+                nextSlave(masterTimestamp);
+            }
+
+            // we have to set the `isMasterHasNextPending` only now since `nextSlave()` may throw DataUnavailableException
+            // and in such case we do not want to call `masterCursor.hasNext()` during the next call to `this.hasNext()`.
+            // if we are here then it's clear nestSlave() did not throw DataUnavailableException.
+            isMasterHasNextPending = true;
+
+            boolean hasSlave = record.hasSlave();
+            if (!hasSlave) {
+                return true;
```

Comment:
Why do we have this early exit? It's not very clear under which conditions we may hit it, so a clarifying comment would help.